Read in nonuniform grid

anne-glerum opened the pull request:
This way we can also read in data that is not equally spaced in a direction, for example crust1.0.

bangerth commented on include/aspect/utilities.h:
```diff
@@ -332,12 +332,12 @@ namespace aspect
          * Interpolation functions to access the data.
          */
-        std::vector<Functions::InterpolatedUniformGridData<dim> *> data;
+        std::vector<Functions::InterpolatedTensorProductGridData<dim> *> data;
```
Can you instead just make it a `std::vector<Function<dim>*>`, for reasons that will become clear below?

bangerth replied:
(Note: `Function<dim>` is a base class for both the uniform and tensor product classes.)

bangerth commented on source/utilities.cc:
```diff
@@ -713,31 +713,40 @@ namespace aspect
-                                                                     table_intervals,
-                                                                     data_tables[dim+i]);
+          data[i] = new Functions::InterpolatedTensorProductGridData<dim> (coordinate_values,
+                                                                           data_tables[dim+i]);
```
The latter of these two classes is significantly more expensive than the former. I would very much like it if this could be changed in the following way:
```
  if (given points are equidistant)
  {
    grid_extent = ...;
    data[i] = new Functions::InterpolatedUniformGridData<dim> (grid_extent,
                                                                     table_intervals,
                                                                     data_tables[dim+i]);
  }
  else
  {
    coordinate_values = ...;
    data[i] = new Functions::InterpolatedTensorProductGridData<dim> (coordinate_values,
                                                                           data_tables[dim+i]);
  }
```
Do you think that would be possible?

anne-glerum replied:
Good point! Should the user specify whether the points are equidistant or not? Or do you want to deduce it from the data file?

bangerth replied:
Good question. I'll gladly leave that up to you. If you want to deduce it automatically, you'll need to have some tolerance. If you want to let the user specify it, you'll need to assert that the points really are equidistant.

gassmoeller commented:
/run-tests

gassmoeller commented on source/utilities.cc:
```diff
@@ -713,31 +713,90 @@ namespace aspect
+          coordinate_values[i].push_back(temp_coord);
+
+          // The grid spacing
+          double first_grid_spacing = 0, grid_spacing = 0;
```
We usually split these declarations into two lines, and I do not think you need to initialize these variables (this is my mandatory comment, which proofs that I have looked through the code :smile:).

gassmoeller commented:
Very nice and useful addition. I assume you have tested the result? Could you maybe take one of the ascii_data tests and modify them to use a non-uniform grid? One test and one changed file should be enough.

anne-glerum commented:
@gassmoeller I've incorporated your comments and added a test. I've tested the result by changing existing test data files, now I'll just do one more test with a data file I've generated. I'll let you know :)

anne-glerum commented:
@gassmoeller I've changed the tolerance used for comparing the grid spacing and added output when the data is found to be nonuniform.

gassmoeller commented on source/utilities.cc:
```diff
@@ -768,8 +769,9 @@ namespace aspect
-                      std::numeric_limits<double>::epsilon()*std::abs(grid_spacing+first_grid_spacing)*2)
+                  // Compare current grid spacing with first grid spacing,
+                  // taking into account roundoff of the read-in coordinates
+                  if (grid_spacing / first_grid_spacing < 0.995 || grid_spacing / first_grid_spacing > 1.005)
```
I think the previous version of this check was more intuitive. Is your current one not equivalent to replacing `std::numeric_limits<double>::epsilon()` by 0.005? I would be favouring that if you need a larger tolerance.

gassmoeller commented:
Thanks! Apart from my one stylistic comment this is ready to go. Care to add an entry in changes.h about the new possibility?
